src/devices/Board/Board.cs

```diff
@@ -0,0 +1,407 @@
+                        if (reservations.Count == 0)
+                        {
+                            _pinReservations.Remove(logicalPin);
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Cannot release Pin {pinNumber}, because it is not reserved.");
+                }
+            }
+        }
+
+        public abstract PinUsage DetermineCurrentPinUsage(int pinNumber);
+
+        /// <summary>
+        /// Override this method if something special needs to be done to use the pin for the given device.
+        /// Many devices support multiple functions per Pin, but not at the same time, so that some kind of
+        /// multiplexer needs to be set accordingly.
+        /// </summary>
+        /// <param name="pinNumber">The logical pin number to use.</param>
+        /// <param name="usage">The intended usage</param>
+        protected virtual void ActivatePinMode(int pinNumber, PinUsage usage)
+        {
+        }
+
+        public virtual void Initialize()
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(ToString());
+            }
+
+            _initialized = true;
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            _disposed = true;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private sealed class PinReservation
+        {
+            public PinReservation(int pin, PinUsage usage, object owner)
+            {
+                Pin = pin;
+                Usage = usage;
+                Owner = owner;
+            }
+
+            public int Pin { get; }
+            public PinUsage Usage { get; }
+
+            /// <summary>
+            /// Component that owns the pin (used mainly for debugging)
+            /// </summary>
+            public object Owner { get; }
+        }
+
+        public virtual GpioController CreateGpioController(int[] pinAssignment = null)
+        {
+            return CreateGpioController(pinAssignment, DefaultPinNumberingScheme);
+        }
+
+        public abstract GpioController CreateGpioController(int[] pinAssignment, PinNumberingScheme pinNumberingScheme);
+
+        /// <summary>
+        /// This method shall be overriden by clients, providing just the basic interface to an I2cDevice.
+        /// </summary>
+        /// <param name="connectionSettings">I2C Connection settings</param>
+        /// <param name="pinAssignment">Pins assigned to the device</param>
+        /// <returns>An I2cDevice connection</returns>
+        protected abstract I2cDevice CreateSimpleI2cDevice(I2cConnectionSettings connectionSettings, int[] pinAssignment);
+
+        public I2cDevice CreateI2cDevice(I2cConnectionSettings connectionSettings, int[] pinAssignment, PinNumberingScheme pinNumberingScheme)
```

Comment:
this interface will always create a device manager even if the same connection settings are used. it would be more interesting if there was the option here to track the connections setting used so far, that would make the board something that can be inspected like the pitop boards and plates in this [repo](https://github.com/colombod/pi-top) .

Reply:
So you would want to make sure that each device (here: every instance that uses the same address) can be created only once? Could indeed further reduce usage errors, since most devices won't like it if they're being used i.e. from multiple bindings at the same time.

Reply:
Yes you are spot on. In my simple implementation reusing a connection settings yield to the already built device. the interesting part is that the device managers can then auto remove themselves if disposed. Do you think this could imply the need for some sort of reference counting? there are some useful disposables classes in rx.Net that could be very useful like refcount disposables, serial disposables and disposable factories.

Reply:
I'll consider this for later, I think. Adding a test that no two I2cDevice instances with the same ConnectionSettings can be done without changing the API later.